src/prototype.rs

```diff
@@ -1546,51 +1594,223 @@ fn make_code_span(tree: &mut Tree<Item>, s: &str, open: usize, close: usize) {
+        if c == open {
+            if open == '(' {
+                nest += 1;
+            } else {
+                return Some(title);
+            }
+        }
+        if open == '(' && c == ')' {
+            if nest == 0 {
+                return Some(title);
+            } else {
+                nest -= 1;
+            }
+        }
+        if c == '\\' {
+            let c = scanner.next_char()?;
+            if !(c <= '\x7f' && is_ascii_punctuation(c as u8)) {
+                title.push('\\');
+            }
+            title.push(c);
+        } else {
+            title.push(c);
+        }
+    }
+    None
+}
+
+fn scan_inline_link(scanner: &mut InlineScanner) -> Option<(String, String)> {
+    if !scanner.scan_ch(b'(') {
+        return None;
+    }
+    scanner.scan_while(is_ascii_whitespace);
+    let url = scan_link_destination(scanner)?;
+    let mut title = String::new();
+    let save = scanner.clone();
+    if scanner.scan_while(is_ascii_whitespace) > 0 {
+        if let Some(t) = scan_link_title(scanner) {
+            title = t;
+            scanner.scan_while(is_ascii_whitespace);
+        } else {
+            *scanner = save;
+        }
+    }
+    if !scanner.scan_ch(b')') {
+        return None;
+    }
+    Some((url, title))
+}
+
+// TODO: if it's just node, get rid of struct. But I think there probably will be
+// state related to normalization.
+struct LinkStackEl {
+    node: usize,
+}
+
+/// Handle inline HTML, code spans, and links.
 ///
 /// This function handles both inline HTML and code spans, because they have
-/// the same precedence. (Maybe rename?)
-fn handle_inline_html(tree: &mut Tree<Item>, s: &str) {
+/// the same precedence. It also handles links, even though they have lower
+/// precedence, because the URL of links must not be processed.
+fn handle_inline_pass1(tree: &mut Tree<Item>, s: &str) {
+    let mut link_stack = Vec::new();
     let mut cur = tree.cur;
+    let mut prev = NIL;
     while cur != NIL {
-        if let ItemBody::MaybeHtml = tree.nodes[cur].item.body {
-            let maybe_html = {
-                let next = tree.nodes[cur].next;
-                let mut scanner = InlineScanner::new(tree, s, next);
-                if scan_inline_html(&mut scanner) {
-                    Some(scanner.to_node_and_ix())
-                } else {
-                    None
-                }
-            };
-            if let Some((node, ix)) = maybe_html {
-                // TODO: this logic isn't right if the replaced chain has
-                // tricky stuff (skipped containers, replaced nulls).
-                tree.nodes[cur].item.body = ItemBody::InlineHtml;
-                tree.nodes[cur].item.end = ix;
-                tree.nodes[cur].next = node;
-                cur = node;
-                if cur != NIL {
-                    tree.nodes[cur].item.start = ix;
+        match tree.nodes[cur].item.body {
+            ItemBody::MaybeHtml => {
```

Comment:
We should also handle autolinks in this block at some point right? Add a TODO?

Reply:
I'm using the spec tests as the main engine driving development. So we know we're going to need autolinks without having an explicit TODO. I do use TODO for things that might not be covered by the spec tests or temporary code that I know needs to be reworked.